src/analysis/check_path.ml

```diff
@@ -0,0 +1,146 @@
+      let formated_start_node = format_node (get_source (Path.start path)) in
+      let formated_rest_nodes = List.map (Seq.to_list @@ Path.edges path) ~f:(fun e -> format_node (get_destination e)) in
+      let formated_full_path = "(" ^ formated_start_node ^ ", " ^ (String.concat ~sep:", " formated_rest_nodes) ^ ")" in
+      formated_full_path
+    end
+
+let find_subfunction_name program name =
+  Term.enum sub_t program
+  |> Seq.find_map ~f:(fun s -> Option.some_if (contains_substring name (Sub.name s)) (Term.tid s))
+
+let get_tids_from_cwe_hit (cwe_hit: Log_utils.CweWarning.t) =
+  cwe_hit.tids
+
+let reaches cg callee target =
+  Graphlib.is_reachable (module CG) cg callee target
+
+(* ignores indirect calls and jumps as well as return statements and interupts *)
+let callsites cg target sub =
+  Term.enum blk_t sub |>
+  Seq.concat_map ~f:(fun blk ->
+      Term.enum jmp_t blk |> Seq.filter_map ~f:(fun j ->
+          match Jmp.kind j with
+          | Goto _ | Ret _ | Int (_,_) -> None
+          | Call destination -> begin match Call.target destination with
+            | Direct tid when reaches cg tid target -> Some (Term.tid blk)
+            | _ -> None
+                                end))
+
+let verify source destination program : proof option =
+  let cg = Program.to_graph program in
+  match Graphlib.shortest_path (module CG) cg source destination with
+  | Some path -> Some (Calls path)
+  | None ->
+     Term.enum sub_t program |> Seq.find_map ~f:(fun sub ->
+        let g = Sub.to_graph sub in
+        Seq.find_map (callsites cg source sub) ~f:(fun sc ->
+            Seq.find_map (callsites cg destination sub) ~f:(fun dc ->
+                if Tid.equal sc dc then None
+                else Graphlib.shortest_path (module CFG) g sc dc))) |>
+    Option.map ~f:(fun p -> Sites p)
+
+let get_fst_tid_from_cwe_hit (cwe_hit: Log_utils.CweWarning.t) =
+  match cwe_hit.tids with
+  | [] -> None
+  | hd :: _ -> Some (Bap.Std.Tid.from_string_exn hd)
+
+let cwe_hit_fst_addr cwe_hit =
+   match get_tids_from_cwe_hit cwe_hit with
+  | [] -> Bap.Std.Tid.from_string_exn "0x00"
+  | hd :: _ -> Bap.Std.Tid.from_string_exn hd
+
+let block_has_callsite blk t =
+  Term.enum jmp_t blk |>
+    Seq.exists ~f:(fun j ->
+        match Jmp.kind j with
+        | Goto _ | Ret _ | Int (_,_) -> false 
+        | Call destination -> begin match Call.target destination with
+                              | Direct tid -> tid = t
+                              | _ -> false
+                              end)
+
+let collect_callsites program t =
+  Term.enum sub_t program
+  |> Seq.map ~f:(fun s -> if Term.enum blk_t s |>
+                               Seq.exists ~f:(fun b -> block_has_callsite b t) then Some s else None)
+  |> Seq.filter ~f:(fun s -> match s with
+                             | None -> false
+                             | _ -> true)
```

Comment:
Isn't there a filter_map method for Seq.t for combining both?

Reply:
Done.